### ISSUE TITLE
Producer Consumer Model For Cook and SharedDataWaiter

### ISSUE THREAD

**Issue by zhangtdavid** [usc-csci201-fall2013/simcity201#36]
The way the Producer-Consumer model is set up according to http://www-scf.usc.edu/~csci201/team/shared-data.html , the consumer will stay in the remove function until an item is added by a producer. In the context of our restaurant, the Cook will stay in the remove function of the RevolvingStand until an Order is added by a Waiter.

The problem I'm having with this is that the Cook can't run its scheduler to do other tasks like "order food from the market" or "take food off the grill" when the RevolvingStand is empty unless an Order is added to the RevolvingStand. I'm thinking you could have the remove function run the cook's scheduler while it's waiting, but the RevolvingStand has no reference to the Cook.

I'm thinking to pass in a reference of whatever agent is calling remove on RevolvingStand into the remove function of RevolvingStand, and in the waiting part of the remove function just call the scheduler at regular intervals. Is this an acceptable solution?

**Comment by xli438** [comment]
Is this new feature of restaurant due in v1 or v2?

**Comment by Valakor** [comment]
I'm pretty sure there's a method similar to a semaphore's mySem.tryAcquire() that attempts to remove something, but simply moves on if it isn't able to. This way the cook would attempt to take something from the revolving stand but simply move on if he can't.

**Comment by EliteSoba** [comment]
Maybe I'm not understanding the issue well because I just woke up, but is there anything wrong with the cook simply checking if there's anything on the stand and moving on if there's not?

**Comment by crowley1958** [comment]
There is nothing wrong with the Cook moving on if there is nothing on the stand.

**Comment by dwilczyn** [comment]
Same for the waiter, if there is no room, just move on to something else. Just add some methods that check the number of entries currently in the stand. The cook and waiters CANNOT block just because the stand is empty or full, respectively.

These are individual points, let's say this feature is due in v2.

**Comment by zhangtdavid** [comment]
Cool! So one more question: how does the cook know if something is added to the stand? Should the stand just message the cook, since the waiter isn't allowed to do that anymore? It seems just like an extra step between messages in that case, which doesn't seem very novel.

**Comment by dwilczyn** [comment]
No, the stand doesn't send messages. As you point out, that would make the whole exercise pointless. This was discussed in class. In case you missed it, try and solve it for yourself.
